lib/jsdom/living/helpers/http-request.js

```diff
@@ -0,0 +1,272 @@
+    this._ending = false;
+    this._redirectCount = 0;
+    this._requestBodyLength = 0;
+    this._requestBodyBuffers = [];
+    this.response = Object.create(null);
+    this._performRequest();
+  }
+
+  abort() {
+    abortRequest(this._currentRequest);
+    this.emit("abort");
+    this.removeAllListeners();
+  }
+
+  getCookie() {
+    let cookies;
+    if (this._redirectCount === 0) {
+      this.originalCookieHeader = this.getHeader("Cookie");
+    }
+    const jar = this._options.cookieJar;
+    if (jar && typeof jar.getCookieStringSync === "function") {
+      cookies = jar.getCookieStringSync(this._currentUrl);
+    } else {
+      this._areCookiesDisabled = true;
+    }
+    if (cookies && cookies.length) {
+      if (this.originalCookieHeader) {
+        this.setHeader("Cookie", this.originalCookieHeader + "; " + cookies);
+      } else {
+        this.setHeader("Cookie", cookies);
+      }
+    }
+  }
+
+  setCookies(res) {
+    const self = this;
+    const jar = self._options.cookieJar;
+    const cookies = res.headers["set-cookie"];
+    if (!this._areCookiesDisabled && jar && typeof jar.setCookieSync === "function" && Array.isArray(cookies)) {
+      try {
+        cookies.forEach(cookie => jar.setCookieSync(cookie, self._currentUrl, { ignoreError: true }));
+      } catch (e) {
+        this.emit("error", e);
+      }
+    }
+  }
+
+  write(data, encoding) {
+    if (this._ending) {
+      const writeAfterEndError = new Error("Writing data after ClientRequest.end() is called is not allowed");
+      writeAfterEndError.code = "ERR_STREAM_WRITE_AFTER_END";
+      throw writeAfterEndError;
+    }
+    if (!(typeof data === "string" || (typeof data === "object" && ("length" in data)))) {
+      throw new TypeError("data should be a string, Buffer or Uint8Array");
+    }
+    if (data.length > 0) {
+      this._requestBodyLength += data.length;
+      this._requestBodyBuffers.push({ data, encoding });
+      this._currentRequest.write(data, encoding);
+    }
+  }
+
+  end() {
+    this.emit("request", this._currentRequest);
+    this._ended = this._ending = true;
+    this._currentRequest.end();
+  }
+
+  setHeader(name, value) {
+    this._options.headers[name] = value;
+    this._currentRequest.setHeader(name, value);
+  }
+
+  removeHeader(name) {
+    delete this._options.headers[name];
+    this._currentRequest.removeHeader(name);
+  }
+
+  toJSON() {
+    const uri = new URL(this._currentUrl);
+    const { method, headers } = this._options;
+    return { uri, method, headers };
+  }
+
+  _performRequest() {
+    const urlOptions = new URL(this._currentUrl);
+    const { protocol: scheme } = urlOptions;
+    this._options.agent = this._options.agents[scheme.substr(0, scheme.length - 1)];
+    const self = this;
+    function onNative(response) {
+      self._processResponse(response);
+    }
+    const { request } = scheme === "https:" ? https : http;
+    const message = request(this._currentUrl, this._options, onNative);
```

Comment:
Why is this file using http/https.request directly, and the other file using node-fetch? It'd be better if they were unified.